tools/monitoring/alert/report.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alert
+
+import (
+	"database/sql"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/knative/test-infra/tools/monitoring/config"
+	"github.com/knative/test-infra/tools/monitoring/log_parser"
+)
+
+const emailTemplate = `In the past %v, 
+The following error pattern reached alerting threshold:
+%s
+
+# occurrences: %v
+jobs affected: %v
+PRs affected: %v
+
+Hint for diagnose & recovery: %s
+
+Error Logs:
+%v
+`
+
+// report stores list of error logs, together with sets of jobs/PRs in those logs
+type report struct {
+	logs []log_parser.ErrorLog
+	jobs []string
+	prs  []int
+}
+
+type mailContent struct {
+	report
+	errorPattern string
+	hint         string
+	window       time.Duration
+}
+
+func (c mailContent) body() string {
+	return fmt.Sprintf(emailTemplate, c.window, c.errorPattern, len(c.logs), c.jobs, c.prs, c.hint, c.logs)
+}
+
+func (c mailContent) subject() string {
+	return fmt.Sprintf("Error pattern reached alerting threshold: %s", c.errorPattern)
+}
+
+// GetErrorLogs returns all jobs stored in ErrorLogs table within the time window
+func GetErrorLogs(s *config.SelectedConfig, errorPattern string, db *sql.DB) ([]log_parser.ErrorLog, error) {
+	var result []log_parser.ErrorLog
+
+	// the timestamp we want to start collecting logs
+	startTime := time.Now().Add(s.Duration())
+
+	rows, err := db.Query(`
+		SELECT ErrorPattern, ErrorMsg, JobName, PRNumber, BuildLogURL, TimeStamp
+		FROM ErrorLogs
+		WHERE ErrorPattern=? and TimeStamp > ?`,
+		errorPattern, startTime)
+
+	if err != nil {
+		return result, err
+	}
+
+	for rows.Next() {
+		entry := log_parser.ErrorLog{}
+		err = rows.Scan(&entry.Pattern, &entry.Msg, &entry.JobName, &entry.PRNumber, &entry.BuildLogURL, &entry.TimeStamp)
+		if err != nil {
+			return result, err
+		}
+		result = append(result, entry)
+	}
+
+	return result, nil
+}
+
+func newReport(errorLogs []log_parser.ErrorLog) *report {
+	report := report{logs: errorLogs}
+
+	// Use sets to store unique values only
+	jobSet := make(map[string]bool)
+	prSet := make(map[int]bool)
+	for _, errorLog := range errorLogs {
+		jobSet[errorLog.JobName] = true
+		prSet[errorLog.PRNumber] = true
+	}
+
+	for job := range jobSet {
+		report.jobs = append(report.jobs, job)
+	}
+	sort.Strings(report.jobs)
```

Comment:
any reason to sort these?

Reply:
yes, we are reporting this in the email eventually and a sorted result looks better